### PR TITLE
Refactor and expand AMP dotcomponents picker

### PR DESCRIPTION
**Pull request by MatthewJWalls**
## What does this change?

Refactors the rendering picker into two subpickers for AMP and Article because the checks are diverging enough to be confusing as one source file.

Makes the AMP picker do proper validation against features and whitelist

# KNOWN REGRESSION:

We no longer log out when a hit *may* be supported by dotcom rendering, only when it *actually is* rendered using dotcomponents.

I will improve the logging under [this separate PR](https://github.com/guardian/frontend/pull/20872) which already is heavily modifying the logging in the validator.

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [ ] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [ ] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [ ] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: article/app/services/dotcomponents/ArticlePicker.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+  def isSupportedType(page: PageWithStoryPackage): Boolean = {
+    page match {
+      case a: ArticlePage => true
+      case _ => false
+    }
+  }
+
+  def hasOnlySupportedElements(page: PageWithStoryPackage): Boolean = {
+    // See: https://github.com/guardian/dotcom-rendering/blob/master/packages/frontend/amp/components/lib/Elements.tsx
```
Comment (SiAdcock) [inline article/app/services/dotcomponents/ArticlePicker.scala]: Should this link to:
https://github.com/guardian/dotcom-rendering/blob/master/packages/frontend/web/components/lib/ArticleRenderer.tsx
?

File: article/app/services/dotcomponents/ArticlePicker.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    def unsupportedElement(blockElement: BlockElement) = blockElement match {
+      case _: TextBlockElement => false
+      case _ => true
```
Comment (SiAdcock) [inline article/app/services/dotcomponents/ArticlePicker.scala]: It looks like we support `ImageBlockElement` on web also

**Comment by SiAdcock** [comment]
This is a nice refactor, and I like the certainty around the new logging ✨ 

Perhaps not the right PR for this, but would it be possible to send a shadow request to `dotcom-rendering` every time frontend receives a request, just to see how `dotcom-rendering` handles the load? 

I guess it's likely `dotcom-rendering` will just hit validation errors before it even gets around to rendering, so maybe this would not be a useful load test.

**Comment by MatthewJWalls** [comment]
>Perhaps not the right PR for this, but would it be possible to send a shadow request to dotcom-rendering every time frontend receives a request, just to see how dotcom-rendering handles the load?

Yep I think we can do this but I'd like to use a different PR for this functionality

File: article/app/services/dotcomponents/ArticlePicker.scala (hunk 2 above)
Comment (MatthewJWalls) [reply article/app/services/dotcomponents/ArticlePicker.scala]: +1

File: article/app/services/dotcomponents/ArticlePicker.scala (hunk 1 above)
Comment (MatthewJWalls) [reply article/app/services/dotcomponents/ArticlePicker.scala]: +1

**Review by SiAdcock — APPROVED** [review]

**Pull request merged**